File: libs/moqtail-rs/src/model/control/publish_blocked.rs
// Copyright 2026 The MOQtail Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use super::constant::ControlMessageType;
use super::control_message::ControlMessageTrait;
use crate::model::common::tuple::{Tuple, TupleField};
use crate::model::common::varint::{BufMutVarIntExt, BufVarIntExt};
use crate::model::error::ParseError;
use bytes::{Buf, BufMut, Bytes, BytesMut};

/// PUBLISH_BLOCKED (0xF): a publisher tells the peer it cannot send a PUBLISH to
/// start a subscription for a track under a SUBSCRIBE_TRACKS namespace because it
/// is blocked by the peer's bidirectional stream limit. Since it always answers a
/// SUBSCRIBE_TRACKS, only the namespace suffix after the prefix is carried.
#[derive(Debug, PartialEq, Clone)]
pub struct PublishBlocked {
  pub track_namespace_suffix: Tuple,
  pub track_name: TupleField,
}

impl PublishBlocked {
  pub fn new(track_namespace_suffix: Tuple, track_name: TupleField) -> Self {
    Self {
      track_namespace_suffix,
      track_name,
    }
  }
}

impl ControlMessageTrait for PublishBlocked {
  fn serialize(&self) -> Result<Bytes, ParseError> {
    let mut buf = BytesMut::new();
    buf.put_vi(ControlMessageType::PublishBlocked)?;

    let mut payload = BytesMut::new();
    payload.extend_from_slice(&self.track_namespace_suffix.serialize()?);
    payload.put_vi(self.track_name.len())?;
    payload.extend_from_slice(self.track_name.as_bytes());

    let payload_len: u16 = payload
      .len()
      .try_into()
      .map_err(|e: std::num::TryFromIntError| ParseError::CastingError {
        context: "PublishBlocked::serialize(payload_length)",
        from_type: "usize",
        to_type: "u16",
        details: e.to_string(),
      })?;
    buf.put_u16(payload_len);
    buf.extend_from_slice(&payload);

    Ok(buf.freeze())
  }

  fn parse_payload(payload: &mut Bytes) -> Result<Box<Self>, ParseError> {
    let track_namespace_suffix = Tuple::deserialize(payload)?;
    let name_len = payload.get_vi()? as usize;
    if payload.remaining() < name_len {
      return Err(ParseError::NotEnoughBytes {
        context: "PublishBlocked::parse_payload(track_name)",
        needed: name_len,
        available: payload.remaining(),
      });
    }
    let track_name = TupleField::new(payload.copy_to_bytes(name_len));

    Ok(Box::new(PublishBlocked {
      track_namespace_suffix,
      track_name,
    }))
  }

  fn get_type(&self) -> ControlMessageType {
    ControlMessageType::PublishBlocked
  }
}

#[cfg(test)]
mod tests {
  use super::*;

  fn sample() -> PublishBlocked {
    PublishBlocked::new(
      Tuple::from_utf8_path("room1/audio"),
      TupleField::from_utf8("track-42"),
    )
  }

  #[test]
  fn test_roundtrip() {
    let msg = sample();
    let mut buf = msg.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::PublishBlocked as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());
    let deserialized = PublishBlocked::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, msg);
    assert!(!buf.has_remaining());
  }

  #[test]
  fn test_excess_roundtrip() {
    let msg = sample();
    let serialized = msg.serialize().unwrap();
    let mut excess = BytesMut::new();
    excess.extend_from_slice(&serialized);
    excess.extend_from_slice(&[9u8, 1u8, 1u8]);
    let mut buf = excess.freeze();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::PublishBlocked as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining() - 3);
    let deserialized = PublishBlocked::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, msg);
    assert_eq!(buf.chunk(), &[9u8, 1u8, 1u8]);
  }

  #[test]
  fn test_partial_message() {
    let msg = sample();
    let mut buf = msg.serialize().unwrap();
    let _ = buf.get_vi().unwrap();
    let _ = buf.get_u16();
    let upper = buf.remaining() / 2;
    let mut partial = buf.slice(..upper);
    assert!(PublishBlocked::parse_payload(&mut partial).is_err());
  }
}

File: libs/moqtail-ts/src/model/control/publish_blocked.ts
/**
 * Copyright 2026 The MOQtail Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { BaseByteBuffer, ByteBuffer, FrozenByteBuffer } from '../common/byte_buffer'
import { Tuple } from '../common/tuple'
import { ControlMessageType } from './constant'
import { LengthExceedsMaxError } from '../error/error'

/**
 * @public
 * PUBLISH_BLOCKED (0xF): a publisher tells the peer it cannot send the PUBLISH that
 * would start a subscription to a track under a SUBSCRIBE_TRACKS namespace, because it
 * is blocked by the peer's bidirectional stream limit. It MUST NOT send PUBLISH for
 * that track until the limit lifts (§10.20).
 *
 * Sent on the SUBSCRIBE_TRACKS response stream, so only the namespace suffix after that
 * subscription's prefix is carried.
 */
export class PublishBlocked {
  constructor(
    public readonly trackNamespaceSuffix: Tuple,
    public readonly trackName: Uint8Array,
  ) {}

  getType(): ControlMessageType {
    return ControlMessageType.PublishBlocked
  }

  serialize(): FrozenByteBuffer {
    const buf = new ByteBuffer()
    buf.putVI(ControlMessageType.PublishBlocked)
    const payload = new ByteBuffer()
    payload.putTuple(this.trackNamespaceSuffix)
    payload.putLengthPrefixedBytes(this.trackName)
    const payloadBytes = payload.toUint8Array()
    if (payloadBytes.length > 0xffff) {
      throw new LengthExceedsMaxError('PublishBlocked::serialize(payloadBytes.length)', 0xffff, payloadBytes.length)
    }
    buf.putU16(payloadBytes.length)
    buf.putBytes(payloadBytes)
    return buf.freeze()
  }

  static parsePayload(buf: BaseByteBuffer): PublishBlocked {
    const trackNamespaceSuffix = buf.getTuple()
    const trackName = buf.getLengthPrefixedBytes()
    return new PublishBlocked(trackNamespaceSuffix, trackName)
  }
}

if (import.meta.vitest) {
  const { describe, test, expect } = import.meta.vitest

  describe('PublishBlocked', () => {
    const sample = () => new PublishBlocked(Tuple.fromUtf8Path('room1/audio'), new TextEncoder().encode('track-42'))

    test('roundtrip', () => {
      const msg = sample()
      const frozen = msg.serialize()
      const msgType = frozen.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.PublishBlocked))
      expect(msgType).toBe(0x0fn)
      const msgLength = frozen.getU16()
      expect(msgLength).toBe(frozen.remaining)
      const deserialized = PublishBlocked.parsePayload(frozen)
      expect(deserialized.trackNamespaceSuffix.equals(msg.trackNamespaceSuffix)).toBe(true)
      expect(deserialized.trackName).toEqual(msg.trackName)
      expect(frozen.remaining).toBe(0)
    })

    test('excess roundtrip', () => {
      const msg = sample()
      const buf = new ByteBuffer()
      buf.putBytes(msg.serialize().toUint8Array())
      buf.putBytes(new Uint8Array([9, 1, 1]))
      const frozen = buf.freeze()
      expect(frozen.getVI()).toBe(BigInt(ControlMessageType.PublishBlocked))
      const msgLength = frozen.getU16()
      expect(msgLength).toBe(frozen.remaining - 3)
      const deserialized = PublishBlocked.parsePayload(frozen)
      expect(deserialized.trackNamespaceSuffix.equals(msg.trackNamespaceSuffix)).toBe(true)
      expect(deserialized.trackName).toEqual(msg.trackName)
      expect(Array.from(frozen.getBytes(3))).toEqual([9, 1, 1])
    })

    test('partial message', () => {
      const serialized = sample().serialize().toUint8Array()
      const frozen = new FrozenByteBuffer(serialized.slice(0, Math.floor(serialized.length / 2)))
      expect(() => {
        frozen.getVI()
        frozen.getU16()
        PublishBlocked.parsePayload(frozen)
      }).toThrow()
    })

    test('a zero-element suffix round-trips', () => {
      const msg = new PublishBlocked(new Tuple(), new TextEncoder().encode('track-42'))
      const frozen = msg.serialize()
      frozen.getVI()
      frozen.getU16()
      const deserialized = PublishBlocked.parsePayload(frozen)
      expect(deserialized.trackNamespaceSuffix.fields.length).toBe(0)
      expect(deserialized.trackName).toEqual(msg.trackName)
    })
  })
}
